src/lib/parser.rs

```diff
@@ -0,0 +1,185 @@
+use lrlex::Lexeme;
+use lrtable::{Action, Grammar, RIdx, StateTable, Symbol};
+
+pub enum Node {
+    Terminal{lexeme: Lexeme},
+    Nonterminal{rule_idx: RIdx, nodes: Vec<Node>}
+}
+
+#[derive(Debug)]
+pub struct ParseError;
+
+impl Node {
+    /// Return a pretty-printed version of this node. In general, you should call this with
+    /// `indent` set to 0.
+    pub fn pp(&self, grm: &Grammar, input: &str, indent: usize) -> String {
+        let mut s = String::new();
+        for _ in 0..indent {
+            s.push_str(" ");
+        }
+        match self {
+            &Node::Terminal{lexeme: Lexeme{tok_id, start, len}} => {
+                s.push_str(&format!("{} {}\n", grm.terminal_names.get(tok_id).unwrap(), &input[start..start + len]));
+            }
+            &Node::Nonterminal{rule_idx, ref nodes} => {
+                s.push_str(&format!("{}\n", grm.nonterminal_names.get(rule_idx).unwrap()));
+                for n in nodes.iter() {
+                    s.push_str(&n.pp(&grm, &input, indent + 1));
+                };
+            }
+        }
+        s
+    }
+}
+
+
+/// Parse the lexemes into a parse tree.
+pub fn parse(grm: &Grammar, stable: &StateTable, lexemes: &Vec<Lexeme>) -> Result<Node, ParseError> {
+    let mut lexemes_iter = lexemes.iter();
+    // Instead of having a single stack, which we'd then have to invent a new struct / tuple for,
+    // it's easiest to split the parse and parse tree stack into two.
+    let mut pstack = vec![0]; // Parse stack
+    let mut tstack: Vec<Node> = Vec::new(); // Parse tree stack
```

Comment:
It could be but then we'd have to make a struct along the lines of:
```
struct StackEntry {
    tree: Option<Node>
    state: usize
}
```
and then do some clever stuff in Reduce to copy the (non-null) trees out. I wasn't convinced it was worth it.